# 01 — Data Preparation

The LendingClub dataset contains ~2.2 million loan records from 2007 to 2018. Before any analysis, we need to understand the structure, quality, and limitations of the data.

This notebook covers:
- Dataset structure and column overview
- Identifying columns relevant to vintage analysis
- Missing values and low variance assessment
- Duplicated rows

In [1]:
import pandas as pd

pd.set_option('display.max_columns', None)

In [2]:
%%time
file = r'..\data\raw\accepted_2007_to_2018q4\accepted_2007_to_2018Q4.csv'
df_raw = pd.read_csv(file)
;

<timed exec>:2: DtypeWarning: Columns (0,19,49,59,118,129,130,131,134,135,136,139,145,146,147) have mixed types. Specify dtype option on import or set low_memory=False.


CPU times: total: 39.7 s
Wall time: 48.2 s


In [3]:
df_raw.head(3)

,id,member_id,loan_amnt,funded_amnt,funded_amnt_inv,term,int_rate,installment,grade,sub_grade,emp_title,emp_length,home_ownership,annual_inc,verification_status,issue_d,loan_status,pymnt_plan,url,desc,purpose,title,zip_code,addr_state,dti,delinq_2yrs,earliest_cr_line,fico_range_low,fico_range_high,inq_last_6mths,mths_since_last_delinq,mths_since_last_record,open_acc,pub_rec,revol_bal,revol_util,total_acc,initial_list_status,out_prncp,out_prncp_inv,total_pymnt,total_pymnt_inv,total_rec_prncp,total_rec_int,total_rec_late_fee,recoveries,collection_recovery_fee,last_pymnt_d,last_pymnt_amnt,next_pymnt_d,last_credit_pull_d,last_fico_range_high,last_fico_range_low,collections_12_mths_ex_med,mths_since_last_major_derog,policy_code,application_type,annual_inc_joint,dti_joint,verification_status_joint,acc_now_delinq,tot_coll_amt,tot_cur_bal,open_acc_6m,open_act_il,open_il_12m,open_il_24m,mths_since_rcnt_il,total_bal_il,il_util,open_rv_12m,open_rv_24m,max_bal_bc,all_util,total_rev_hi_lim,inq_fi,total_cu_tl,inq_last_12m,acc_open_past_24mths,avg_cur_bal,bc_open_to_buy,bc_util,chargeoff_within_12_mths,delinq_amnt,mo_sin_old_il_acct,mo_sin_old_rev_tl_op,mo_sin_rcnt_rev_tl_op,mo_sin_rcnt_tl,mort_acc,mths_since_recent_bc,mths_since_recent_bc_dlq,mths_since_recent_inq,mths_since_recent_revol_delinq,num_accts_ever_120_pd,num_actv_bc_tl,num_actv_rev_tl,num_bc_sats,num_bc_tl,num_il_tl,num_op_rev_tl,num_rev_accts,num_rev_tl_bal_gt_0,num_sats,num_tl_120dpd_2m,num_tl_30dpd,num_tl_90g_dpd_24m,num_tl_op_past_12m,pct_tl_nvr_dlq,percent_bc_gt_75,pub_rec_bankruptcies,tax_liens,tot_hi_cred_lim,total_bal_ex_mort,total_bc_limit,total_il_high_credit_limit,revol_bal_joint,sec_app_fico_range_low,sec_app_fico_range_high,sec_app_earliest_cr_line,sec_app_inq_last_6mths,sec_app_mort_acc,sec_app_open_acc,sec_app_revol_util,sec_app_open_act_il,sec_app_num_rev_accts,sec_app_chargeoff_within_12_mths,sec_app_collections_12_mths_ex_med,sec_app_mths_since_last_major_derog,hardship_flag,hardship_type,hardship_reason,hardship_status,deferral_term,hardship_amount,hardship_start_date,hardship_end_date,payment_plan_start_date,hardship_length,hardship_dpd,hardship_loan_status,orig_projected_additional_accrued_interest,hardship_payoff_balance_amount,hardship_last_payment_amount,disbursement_method,debt_settlement_flag,debt_settlement_flag_date,settlement_status,settlement_date,settlement_amount,settlement_percentage,settlement_term
0,68407277,NaN,3600.0,3600.0,3600.0,36 months,13.99,123.03,C,C4,leadman,10+ years,MORTGAGE,55000.0,Not Verified,Dec-2015,Fully Paid,n,https://lendingclub.com/browse/loanDetail.acti...,NaN,debt_consolidation,Debt consolidation,190xx,PA,5.91,0.0,Aug-2003,675.0,679.0,1.0,30.0,NaN,7.0,0.0,2765.0,29.7,13.0,w,0.0,0.0,4421.723917,4421.72,3600.0,821.72,0.0,0.0,0.0,Jan-2019,122.67,NaN,Mar-2019,564.0,560.0,0.0,30.0,1.0,Individual,NaN,NaN,NaN,0.0,722.0,144904.0,2.0,2.0,0.0,1.0,21.0,4981.0,36.0,3.0,3.0,722.0,34.0,9300.0,3.0,1.0,4.0,4.0,20701.0,1506.0,37.2,0.0,0.0,148.0,128.0,3.0,3.0,1.0,4.0,69.0,4.0,69.0,2.0,2.0,4.0,2.0,5.0,3.0,4.0,9.0,4.0,7.0,0.0,0.0,0.0,3.0,76.9,0.0,0.0,0.0,178050.0,7746.0,2400.0,13734.0,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,N,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,Cash,N,NaN,NaN,NaN,NaN,NaN,NaN
1,68355089,NaN,24700.0,24700.0,24700.0,36 months,11.99,820.28,C,C1,Engineer,10+ years,MORTGAGE,65000.0,Not Verified,Dec-2015,Fully Paid,n,https://lendingclub.com/browse/loanDetail.acti...,NaN,small_business,Business,577xx,SD,16.06,1.0,Dec-1999,715.0,719.0,4.0,6.0,NaN,22.0,0.0,21470.0,19.2,38.0,w,0.0,0.0,25679.660000,25679.66,24700.0,979.66,0.0,0.0,0.0,Jun-2016,926.35,NaN,Mar-2019,699.0,695.0,0.0,NaN,1.0,Individual,NaN,NaN,NaN,0.0,0.0,204396.0,1.0,1.0,0.0,1.0,19.0,18005.0,73.0,2.0,3.0,6472.0,29.0,111800.0,0.0,0.0,6.0,4.0,9733.0,57830.0,27.1,0.0,0.0,113.0,192.0,2.0,2.0,4.0,2.0,NaN,0.0,6.0,0.0,5.0,5.0,13.0,17.0,6.0,20.0,27.0,5.0,22.0,0.0,0.0,0.0,2.0,97.4,7.7,0.0,0.0,314017.0,39475.0,79300.0,24667.0,NaN,NaN,NaN,NaN,NaN,NaN,NaN

In [4]:
df_raw.shape

(2260701, 151)

As the Dataset has 151 columns, lets take a look if there are columns completely missing, or filled with just a unique value. If we find it, we can remove before any exploration.

In [5]:
%%time
df_nunique = (
    df_raw
    .nunique()
    .rename_axis('column')
    .to_frame('n_unique')
)

df_nunique.head(10)

CPU times: total: 11.4 s
Wall time: 11.9 s


,n_unique
column,
id,2260701
member_id,0
loan_amnt,1572
funded_amnt,1572
funded_amnt_inv,10057
term,2
int_rate,673
installment,93301
grade,7


In [6]:
%%time
df_missing = (
    df_raw
    .isna()
    .sum()
    .rename('n_missing')
    .rename_axis('column')
    .to_frame()
)

df_missing['pct'] = df_missing['n_missing'] / len(df_raw)

df_missing.head(10)

CPU times: total: 3.3 s
Wall time: 3.3 s


,n_missing,pct
column,,
id,0,0.000000
member_id,2260701,1.000000
loan_amnt,33,0.000015
funded_amnt,33,0.000015
funded_amnt_inv,33,0.000015
term,33,0.000015
int_rate,33,0.000015
installment,33,0.000015
grade,33,0.000015


In [7]:
# Columns with more than 70% missing
high_missing = df_missing.loc[df_missing['pct'] > 0.70].index.tolist()

# Columns with 0 or 1 unique value (no analytical value)
low_variance = df_nunique.loc[df_nunique['n_unique'] <= 1].index.tolist()

# Columns irrelevant by definition (admin/id fields)
admin_cols = ['member_id', 'desc', 'url', 'id']

columns_to_remove = list(set(high_missing + low_variance + admin_cols))

print(f'Columns to remove: {len(columns_to_remove)}')
print(f'Remaining columns: {df_raw.shape[1] - len(columns_to_remove)}')

Columns to remove: 44
Remaining columns: 107


In [8]:
df = df_raw.drop(columns=columns_to_remove).copy()
print(f'Shape after removal: {df.shape}')

Shape after removal: (2260701, 107)


After removing columns with more than 70% missing and those with low variance, the dataset reduced from 151 to 107 columns. However, 107 columns is still a large number to analyse. Fortunately, the dataset's Kaggle's page includes a link to the data dictionary ([see here](https://data.world/jaypeedevlin/lending-club-loan-data-2007-11)). We can use it to identify additional columns that may be unnecessary for our analysis.

For this project, the variables will be separated into three groups: Characteristics related to the loan origination, outcomes after origination, and Bureau information.

In [9]:
data_dict_raw = pd.read_csv(r'..\data\raw\LCDataDictionary.csv')
data_dict = data_dict_raw.loc[data_dict_raw['LoanStatNew'].isin(df.columns.tolist())]

In [10]:
data_dict.head()

,LoanStatNew,Description
0,acc_now_delinq,The number of accounts on which the borrower i...
1,acc_open_past_24mths,Number of trades opened in past 24 months.
2,addr_state,The state provided by the borrower in the loan...
3,all_util,Balance to credit limit on all trades
4,annual_inc,The self-reported annual income provided by th...


In [11]:
cols_selected = [
    # --- Core vintage ---
    'issue_d',                # The month which the loan was funded
    'loan_status',            # outcome variable
    'loan_amnt',              # loan amount
    'grade',                  # risk grade (A to G)
    'sub_grade',              # granular risk grade
    'term',                   # 36 or 60 months
    'int_rate',               # interest rate
    'installment',            # monthly payment
    
    # --- At-origination borrower profile ---
    'annual_inc',             # self-reported income
    'dti',                    # debt-to-income ratio
    'emp_length',             # employment length
    'home_ownership',         # RENT / OWN / MORTGAGE / OTHER
    'addr_state',             # the state provided by the borrower in the loan application
    'purpose',                # loan purpose
    'verification_status',    # indicates if income was verified or not
    'application_type',       # individual or joint loan
    'fico_range_low',         # FICO score at origination
    'fico_range_high',        # FICO score at origination
    'earliest_cr_line',       # credit history length
    'delinq_2yrs',            # past delinquency
    'inq_last_6mths',         # recent credit inquiries
    'open_acc',               # open credit lines
    'pub_rec',                # derogatory records
    'revol_util',             # revolving utilization
    'mths_since_last_delinq', # months since last delinquency
    
    # --- Post-origination outcomes ---
    'last_pymnt_d',           # last month payment was received
    'last_pymnt_amnt',        # last total payment amount received
    'total_pymnt',            # total received to date
    'total_rec_prncp',        # principal received to date
    'total_rec_late_fee',     # Late fees received to date
    'recoveries',             # post charge off gross recovery
    'out_prncp',              # outstanding principal
]

df = df[cols_selected].copy()
print(f'Final shape: {df.shape}')

Final shape: (2260701, 32)


Now that we have selected our columns, let's take a look to see if there are duplicated rows or rows with all missing value.

In [12]:
print(f'Number of duplicated rows: {df.duplicated().sum()}')

Number of duplicated rows: 32


In [13]:
df.drop_duplicates(inplace=True)

In [14]:
print(f'Rows with all missing values: {df.isna().all(axis=1).sum()}')

Rows with all missing values: 1


In [15]:
df.dropna(how='all', inplace=True)

With our data preparation complete, we'll save our dataset for the next step: Exploratory Data Analysis.

In [16]:
output_path = r'..\data\processed\loans_prepared.parquet'
df.to_parquet(output_path, index=False)
print('Done ;)')

Done ;)
